cmd/collector/main.go

```diff
@@ -85,10 +108,27 @@ func main() {
 	r := mux.NewRouter()
 	apiHandler := app.NewAPIHandler(jaegerBatchesHandler, zipkinSpansHandler)
 	apiHandler.RegisterRoutes(r)
+
 	httpPortStr := ":" + strconv.Itoa(*builder.CollectorHTTPPort)
 	recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-	logger.Info("Listening for HTTP traffic", zap.Int("http-port", *builder.CollectorHTTPPort))
-	if err := http.ListenAndServe(httpPortStr, recoveryHandler(r)); err != nil {
-		logger.Fatal("Could not launch service", zap.Error(err))
+	logger.Info("Starting Jaeger Collector HTTP server", zap.Int("http-port", *builder.CollectorHTTPPort))
+
+	go func() {
+		// there's still a chance that a request comes in between the `ready` flag and the server's
+		// actual startup, but this would be a rare situation, as we assume the `ListenAndServe` is fast enough
+		// between allocating a port and serving a request.
+		serverState = http.StatusNoContent
+
+		logger.Fatal("Could not launch service", zap.Error(http.ListenAndServe(httpPortStr, recoveryHandler(r))))
+
+		// if we returned from the call above, it's because we have troubles... the logger.Fatal should panic,
+		// so, this code probably won't be reached, but if we are hung for some reason, we can still notify our
+		// state to an external actor
+		serverState = http.StatusInternalServerError
+	}()
+
+	select {
+	case <-serverChannel:
+		logger.Info("Jaeger Collector is finishing")
```

Comment:
HEADS UP: this change means that stopping the server now ends up with a return code 0. Previously, the return code was 130:

Before this PR:
```
$ ./cmd/collector/collector-linux
{"level":"info","ts":1500553817.5565636,"caller":"collector/main.go:90","msg":"Listening for HTTP traffic","http-port":14268}
^C
$ echo $?
130
```

After this PR:
```
$ ./cmd/collector/collector-linux
{"level":"info","ts":1500553862.3376029,"caller":"collector/main.go:62","msg":"Health Check server started","http-port":15268}
{"level":"info","ts":1500553862.3593538,"caller":"collector/main.go:114","msg":"Starting Jaeger Collector HTTP server","http-port":14268}
^C{"level":"info","ts":1500553864.2055447,"caller":"collector/main.go:132","msg":"Jaeger Collector is finishing"}
$ echo $?
0
```